src/tests/ftest/cart/utest/utest_portnumber.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * (C) Copyright 2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of CaRT testing.
+ */
+
+/*
+ * This code tests whether a provider returns an error if 2 independent
+ * instances attempts to open the same port number.
+ *
+ * Two child process are forked with the same provider information.
+ * The test is setup so that first child opens the port and then
+ * the second child should fail.
+ * Note, synchronization between child process is performed via sleep
+ * function calls.  The sleep time should prevent any problems, but
+ * beware if an issue arise where the results are swapped.
+ */
+/*
+ * Reference jira ticket DAOS-5732 to include socket and verb tests.
+ */
+#define MY_TESTS_NOT_INCLUDED
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <time.h>
+#include <sys/wait.h>
+
+#include <cmocka.h>
+#include <cart/api.h>
+#include "gurt/debug.h"
+
+static void
+run_test_fork(void **state)
+{
+	int		result1 = -1;
+	int		result2 = -1;
+	int		status = -1;
+	int		rc = 0;
+	int		child_result;
+	pid_t		pid1 = 0;
+	pid_t		pid2 = 0;
+	crt_context_t	crt_context = NULL;
+
+	pid1 = fork();
+	/* fork first child process */
+	if (pid1 == 0) {
+		rc = crt_init(NULL, CRT_FLAG_BIT_SERVER);
+		if (rc != 0)
+			exit(10);
+		child_result = crt_context_create(&crt_context);
+		sleep(10);	/* wait for second child */
+		rc = crt_context_destroy(crt_context, false);
+		if (rc != 0)
+			exit(11);
+		rc = crt_finalize();
+		if (rc != 0)
+			exit(12);
+		exit(child_result);
+	}
+
+	/* fork second child process */
+	pid2 = fork();
+	if (pid2 == 0) {
+		sleep(2);  /* wait for first child just in case */
```

Comment:
this is racy.  Why not just allow both to race and then ensure that at least one result is 0 and the other is non-zero?

Reply:
This not a race condition if that is what you meant by racy.  No, you can not just let both of them run and check for non-zero from one of them.  The first fork may run to completion before the second one even starts.  In that case, both of them would return 0.

The code keeps tight synchronization between the 2 threads.  The way to make this code better is to create a semaphore between the 2 code sequence.  However, I was trying to avoid this complication.

Reply:
Created share memory semaphores for synchronization between child processes.